src/EditorFeatures/Core/GoToNextAndPreviousMethodCommandHandler.cs

```diff
@@ -0,0 +1,180 @@
+using Microsoft.CodeAnalysis.LanguageServices;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Text;
+using Microsoft.VisualStudio.Text;
+using Microsoft.VisualStudio.Text.Editor;
+using Microsoft.VisualStudio.Text.Outlining;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.Editor
+{
+    [ExportCommandHandler("Go to next and previous method command handler", ContentTypeNames.RoslynContentType)]
+    class GoToNextAndPreviousMethodCommandHandler : ICommandHandler<GoToNextMethodCommandArgs>, ICommandHandler<GoToPreviousMethodCommandArgs>
+    {
+        private readonly IWaitIndicator _waitIndicator;
+        private readonly IOutliningManagerService _outliningManagerService;
+
+        [ImportingConstructor]
+        public GoToNextAndPreviousMethodCommandHandler(IWaitIndicator waitIndicator, IOutliningManagerService outliningManagerService)
+        {
+            _waitIndicator = waitIndicator;
+            _outliningManagerService = outliningManagerService;
+        }
+
+        public CommandState GetCommandState(GoToNextMethodCommandArgs args, Func<CommandState> nextHandler)
+            => GetCommandState(args.SubjectBuffer, nextHandler);
+
+        public void ExecuteCommand(GoToNextMethodCommandArgs args, Action nextHandler)
+            => ExecuteCommand(nextHandler, args.SubjectBuffer, args.TextView, next: true);
+
+        public CommandState GetCommandState(GoToPreviousMethodCommandArgs args, Func<CommandState> nextHandler)
+            => GetCommandState(args.SubjectBuffer, nextHandler);
+
+        public void ExecuteCommand(GoToPreviousMethodCommandArgs args, Action nextHandler)
+            => ExecuteCommand(nextHandler, args.SubjectBuffer, args.TextView, next: false);
+
+        private static CommandState GetCommandState(ITextBuffer subjectBuffer, Func<CommandState> nextHandler)
+        {
+            var document = subjectBuffer.CurrentSnapshot.GetOpenDocumentInCurrentContextWithChanges();
+            var documentSupportsSuggestionService = document.Project.Solution.Workspace.Services.GetService<IDocumentSupportsSuggestionService>();
+            if (document == null || !document.SupportsSyntaxTree || !documentSupportsSuggestionService.SupportsRefactorings(document))
+            {
+                return nextHandler();
+            }
+
+            return CommandState.Available;
+        }
+
+
+        private void ExecuteCommand(Action nextHandler, ITextBuffer subjectBuffer, ITextView textView, bool next)
+        {
+            var document = subjectBuffer.CurrentSnapshot.GetOpenDocumentInCurrentContextWithChanges();
+            if (document == null || !document.SupportsSyntaxTree)
+            {
+                nextHandler();
+                return;
+            }
+
+            var caretPoint = textView.GetCaretPoint(subjectBuffer);
+            if (!caretPoint.HasValue)
+            {
+                nextHandler();
+                return;
```

Comment:
Should we have also checked this in GetCommandState? Otherwise we're passing the Execute call through to a handler that hasn't declared that they really want it.